File: PennMobile/src/main/java/com/pennapps/labs/pennmobile/adapters/PhoneSaveAdapter.java
package com.pennapps.labs.pennmobile.adapters;

import android.content.Context;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.RadioButton;
import android.widget.TextView;

import com.pennapps.labs.pennmobile.R;
import com.pennapps.labs.pennmobile.classes.Person;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class PhoneSaveAdapter extends ArrayAdapter<Person> {

    private final LayoutInflater inflater;
    private final List<Person> selections;

    public PhoneSaveAdapter(@NonNull Context context, List<Person> contacts, List<Person> s) {
        super(context, R.layout.phone_save_list_item, contacts);
        inflater = LayoutInflater.from(context);
        selections = s;
    }

    @Override
    public View getView(int pos, View view, ViewGroup parent) {
        final Person currentPerson = getItem(pos);
        final PhoneSaveAdapter.ViewHolder holder;
        if (view != null) {
            holder = (PhoneSaveAdapter.ViewHolder) view.getTag();
        } else {
            view = inflater.inflate(R.layout.phone_save_list_item, parent, false);
            holder = new PhoneSaveAdapter.ViewHolder(view);
            view.setTag(holder);
        }

        holder.name.setText(currentPerson.name);
        holder.phone.setText(currentPerson.phone);

        holder.icon.setVisibility(currentPerson.isURL() ? View.GONE: View.VISIBLE);

        holder.radio.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                if (isChecked) {
                    holder.radio.setChecked(true);
                    selections.add(new Person(holder.name.getText().toString(), holder.phone.getText().toString()));
                } else {
                    holder.radio.setChecked(false);
                    for (Person p : selections) {
                        if (p.name.equals(holder.name.getText().toString())) {
                            selections.remove(p);
                            break;
                        }
                    }
                }
            }
        });

        view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                holder.radio.setChecked(!holder.radio.isChecked());
            }
        });
        return view;
    }

    public static class ViewHolder {
        @BindView(R.id.support_name) TextView name;
        @BindView(R.id.support_phone) TextView phone;
        @BindView(R.id.support_phone_icon) ImageView icon;
        @BindView(R.id.phone_save_radiobutton) RadioButton radio;

        public ViewHolder(View view) {
            ButterKnife.bind(this, view);
        }
    }
}

File: PennMobile/src/main/java/com/pennapps/labs/pennmobile/adapters/PhoneSaveAdapter.kt
package com.pennapps.labs.pennmobile.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import com.pennapps.labs.pennmobile.R
import com.pennapps.labs.pennmobile.classes.Contact
import kotlinx.android.synthetic.main.phone_save_list_item.view.*

class PhoneSaveAdapter(context: Context, contacts: List<Contact?>, s: MutableList<Contact>, size: Int) : ArrayAdapter<Contact?>(context, R.layout.phone_save_list_item, contacts) {

    private val inflater: LayoutInflater = LayoutInflater.from(context)
    private val selections: MutableList<Contact> = s
    private val state: BooleanArray = BooleanArray(size)

    override fun getView(pos: Int, view: View?, parent: ViewGroup): View {
        val currentPerson = getItem(pos)
        val view = view ?: inflater.inflate(R.layout.phone_save_list_item, parent, false)

        val nameTv = view.support_name
        val supportPhoneTv = view.support_phone
        val checkbox = view.phone_save_checkbox
        nameTv?.text = currentPerson?.name
        supportPhoneTv?.text = currentPerson?.phone
        view.support_phone_icon?.visibility = if (currentPerson?.isURL == true) View.GONE else View.VISIBLE
        checkbox?.setOnCheckedChangeListener(null)
        checkbox?.isChecked = state[pos]
        checkbox?.setOnCheckedChangeListener { buttonView, isChecked ->
            if (!state[pos]) {
                checkbox.isChecked = true
                selections.add(Contact(nameTv?.text.toString(), supportPhoneTv.text.toString()))
                state[pos] = true
            } else {
                view.phone_save_checkbox?.isChecked = false
                for (p in selections) {
                    if (p.name == nameTv?.text.toString()) {
                        selections.remove(p)
                        break
                    }
                }
                state[pos] = false
            }
        }
        view.setOnClickListener { checkbox.isChecked = !checkbox.isChecked }
        return view
    }

    init {
        for (i in state.indices) {
            state[i] = true
        }
    }
}